File: .github/leetcode-runner/helpers/java/LeetCodeHelper.java
// LeetCode Helper for Java
// Data structure parsing and output formatting

import java.util.*;

public class LeetCodeHelper {

    // ========================================
    // LeetCode Input Parsing
    // ========================================

    /**
     * Parse LeetCode format input
     * Example: "nums = [1,2,3], target = 9" -> ["[1,2,3]", "9"]
     */
    public static List<String> parseLeetCodeInput(String input) {
        List<String> results = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        int bracketDepth = 0;

        for (char c : input.toCharArray()) {
            if (c == '[') {
                bracketDepth++;
                current.append(c);
            } else if (c == ']') {
                bracketDepth--;
                current.append(c);
            } else if (c == ',' && bracketDepth == 0) {
                results.add(extractValue(current.toString()));
                current = new StringBuilder();
            } else {
                current.append(c);
            }
        }
        if (current.length() > 0) {
            results.add(extractValue(current.toString()));
        }
        return results;
    }

    /**
     * Extract value from "name = value" format
     */
    public static String extractValue(String s) {
        s = s.trim();
        int eqIndex = s.indexOf('=');
        if (eqIndex != -1) {
            return s.substring(eqIndex + 1).trim();
        }
        return s;
    }

    // ========================================
    // Parse Functions
    // ========================================

    /**
     * Parse integer array
     */
    public static int[] parseIntArray(String s) {
        s = s.trim().replaceAll("[\\[\\]]", "");
        if (s.isEmpty()) return new int[0];
        String[] parts = s.split(",");
        int[] arr = new int[parts.length];
        for (int i = 0; i < parts.length; i++) {
            arr[i] = Integer.parseInt(parts[i].trim());
        }
        return arr;
    }

    /**
     * Parse 2D integer array (matrix)
     */
    public static int[][] parseIntMatrix(String s) {
        s = s.trim();
        if (s.equals("[]") || s.isEmpty()) return new int[0][];

        List<int[]> rows = new ArrayList<>();
        int depth = 0;
        StringBuilder current = new StringBuilder();

        for (int i = 1; i < s.length() - 1; i++) {
            char c = s.charAt(i);
            if (c == '[') {
                depth++;
                if (depth == 1) current = new StringBuilder();
                current.append(c);
            } else if (c == ']') {
                current.append(c);
                depth--;
                if (depth == 0) {
                    rows.add(parseIntArray(current.toString()));
                }
            } else if (c == ',' && depth == 0) {
                // skip
            } else {
                current.append(c);
            }
        }
        return rows.toArray(new int[0][]);
    }

    /**
     * Parse string array
     */
    public static String[] parseStringArray(String s) {
        s = s.trim().replaceAll("[\\[\\]]", "");
        if (s.isEmpty()) return new String[0];
        String[] parts = s.split(",");
        for (int i = 0; i < parts.length; i++) {
            parts[i] = parts[i].trim().replaceAll("^\"|\"$", "");
        }
        return parts;
    }

    /**
     * Parse nullable integer array (for binary tree)
     */
    public static Integer[] parseIntegerArray(String s) {
        s = s.trim().replaceAll("[\\[\\]]", "");
        if (s.isEmpty()) return new Integer[0];
        String[] parts = s.split(",");
        Integer[] arr = new Integer[parts.length];
        for (int i = 0; i < parts.length; i++) {
            String p = parts[i].trim();
            arr[i] = p.equals("null") ? null : Integer.parseInt(p);
        }
        return arr;
    }

    /**
     * Parse integer
     */
    public static int parseInt(String s) {
        return Integer.parseInt(s.trim());
    }

    /**
     * Parse string (remove quotes)
     */
    public static String parseString(String s) {
        return s.trim().replaceAll("^\"|\"$", "");
    }

    /**
     * Parse boolean
     */
    public static boolean parseBool(String s) {
        s = s.trim().toLowerCase();
        return s.equals("true") || s.equals("1") || s.equals("yes");
    }

    /**
     * Parse with automatic type detection
     */
    public static Object parseInput(String s) {
        s = s.trim();
        if (s.isEmpty()) return "";

        // Array
        if (s.startsWith("[")) {
            // 2D array
            if (s.startsWith("[[")) {
                return parseIntMatrix(s);
            }
            // Check if string array
            if (s.contains("\"")) {
                return parseStringArray(s);
            }
            return parseIntArray(s);
        }

        // Boolean
        String lower = s.toLowerCase();
        if (lower.equals("true") || lower.equals("false")) {
            return parseBool(s);
        }

        // Integer
        try {
            return Integer.parseInt(s);
        } catch (NumberFormatException e) {
            // not an integer
        }

        // Double
        try {
            return Double.parseDouble(s);
        } catch (NumberFormatException e) {
            // not a double
        }

        // String
        return parseString(s);
    }

    // ========================================
    // Linked List Operations
    // ========================================

    /**
     * Convert array to linked list
     */
    public static ListNode arrayToList(int[] arr) {
        if (arr == null || arr.length == 0) return null;
        ListNode head = new ListNode(arr[0]);
        ListNode current = head;
        for (int i = 1; i < arr.length; i++) {
            current.next = new ListNode(arr[i]);
            current = current.next;
        }
        return head;
    }

    /**
     * Convert linked list to array
     */
    public static int[] listToArray(ListNode head) {
        List<Integer> list = new ArrayList<>();
        ListNode current = head;
        while (current != null) {
            list.add(current.val);
            current = current.next;
        }
        return list.stream().mapToInt(i -> i).toArray();
    }

    /**
     * Parse string to linked list
     */
    public static ListNode parseLinkedList(String s) {
        return arrayToList(parseIntArray(s));
    }

    // ========================================
    // Binary Tree Operations
    // ========================================

    /**
     * Convert array to binary tree
     */
    public static TreeNode arrayToTree(Integer[] arr) {
        if (arr == null || arr.length == 0 || arr[0] == null) return null;

        TreeNode root = new TreeNode(arr[0]);
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int i = 1;

        while (!queue.isEmpty() && i < arr.length) {
            TreeNode node = queue.poll();

            // Left child
            if (i < arr.length) {
                if (arr[i] != null) {
                    node.left = new TreeNode(arr[i]);
                    queue.offer(node.left);
                }
                i++;
            }

            // Right child
            if (i < arr.length) {
                if (arr[i] != null) {
                    node.right = new TreeNode(arr[i]);
                    queue.offer(node.right);
                }
                i++;
            }
        }

        return root;
    }

    /**
     * Convert binary tree to array
     */
    public static Integer[] treeToArray(TreeNode root) {
        if (root == null) return new Integer[0];

        List<Integer> result = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            if (node != null) {
                result.add(node.val);
                queue.offer(node.left);
                queue.offer(node.right);
            } else {
                result.add(null);
            }
        }

        // Remove trailing nulls
        while (result.size() > 0 && result.get(result.size() - 1) == null) {
            result.remove(result.size() - 1);
        }

        return result.toArray(new Integer[0]);
    }

    /**
     * Parse string to binary tree
     */
    public static TreeNode parseTree(String s) {
        return arrayToTree(parseIntegerArray(s));
    }

    // ========================================
    // Output Formatting
    // ========================================

    /**
     * Generic output formatter
     */
    public static String formatOutput(Object val) {
        if (val == null) return "null";
        if (val instanceof int[]) {
            return Arrays.toString((int[])val).replaceAll(" ", "");
        } else if (val instanceof int[][]) {
            StringBuilder sb = new StringBuilder("[");
            int[][] matrix = (int[][])val;
            for (int i = 0; i < matrix.length; i++) {
                if (i > 0) sb.append(",");
                sb.append(Arrays.toString(matrix[i]).replaceAll(" ", ""));
            }
            sb.append("]");
            return sb.toString();
        } else if (val instanceof Integer[]) {
            return Arrays.toString((Integer[])val).replaceAll(" ", "");
        } else if (val instanceof String[]) {
            StringBuilder sb = new StringBuilder("[");
            String[] arr = (String[])val;
            for (int i = 0; i < arr.length; i++) {
                if (i > 0) sb.append(",");
                sb.append("\"").append(arr[i]).append("\"");
            }
            sb.append("]");
            return sb.toString();
        } else if (val instanceof ListNode) {
            return Arrays.toString(listToArray((ListNode)val)).replaceAll(" ", "");
        } else if (val instanceof TreeNode) {
            Integer[] arr = treeToArray((TreeNode)val);
            StringBuilder sb = new StringBuilder("[");
            for (int i = 0; i < arr.length; i++) {
                if (i > 0) sb.append(",");
                sb.append(arr[i] == null ? "null" : arr[i]);
            }
            sb.append("]");
            return sb.toString();
        } else if (val instanceof Boolean) {
            return ((Boolean)val) ? "true" : "false";
        } else if (val instanceof List) {
            return val.toString().replaceAll(" ", "");
        }
        return String.valueOf(val);
    }
}

File: .github/leetcode-runner/helpers/swift/LeetCodeHelper.swift
// LeetCode Helper for Swift
// Data structures and parsing functions

import Foundation

// ========================================
// Data Structure Definitions
// ========================================

// ListNode - Linked list node
public class ListNode {
    public var val: Int
    public var next: ListNode?

    public init() {
        self.val = 0
        self.next = nil
    }

    public init(_ val: Int) {
        self.val = val
        self.next = nil
    }

    public init(_ val: Int, _ next: ListNode?) {
        self.val = val
        self.next = next
    }
}

// TreeNode - Binary tree node
public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?

    public init() {
        self.val = 0
        self.left = nil
        self.right = nil
    }

    public init(_ val: Int) {
        self.val = val
        self.left = nil
        self.right = nil
    }

    public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
        self.val = val
        self.left = left
        self.right = right
    }
}

// ========================================
// Parsing Functions
// ========================================

// Parse integer array from string
public func parseIntArray(_ s: String) -> [Int] {
    let trimmed = s.trimmingCharacters(in: .whitespaces)
    if trimmed.isEmpty || trimmed == "[]" {
        return []
    }

    // Try JSON parsing
    if let data = trimmed.data(using: .utf8),
       let arr = try? JSONSerialization.jsonObject(with: data) as? [Int] {
        return arr
    }

    // Fallback: comma-separated
    let cleaned = trimmed
        .replacingOccurrences(of: "[", with: "")
        .replacingOccurrences(of: "]", with: "")

    return cleaned.split(separator: ",").compactMap { Int($0.trimmingCharacters(in: .whitespaces)) }
}

// Parse string array from string
public func parseStringArray(_ s: String) -> [String] {
    let trimmed = s.trimmingCharacters(in: .whitespaces)
    if trimmed.isEmpty || trimmed == "[]" {
        return []
    }

    // Try JSON parsing
    if let data = trimmed.data(using: .utf8),
       let arr = try? JSONSerialization.jsonObject(with: data) as? [String] {
        return arr
    }

    // Fallback: comma-separated
    let cleaned = trimmed
        .replacingOccurrences(of: "[", with: "")
        .replacingOccurrences(of: "]", with: "")

    return cleaned.split(separator: ",").map {
        $0.trimmingCharacters(in: .whitespaces)
          .trimmingCharacters(in: CharacterSet(charactersIn: "\"'"))
    }
}

// Parse nullable integer array (for binary tree)
public func parseNullableIntArray(_ s: String) -> [Int?] {
    let trimmed = s.trimmingCharacters(in: .whitespaces)
    if trimmed.isEmpty || trimmed == "[]" {
        return []
    }

    // Try JSON parsing
    if let data = trimmed.data(using: .utf8),
       let arr = try? JSONSerialization.jsonObject(with: data) as? [Any] {
        return arr.map { item in
            if let num = item as? Int {
                return num
            } else if let num = item as? Double {
                return Int(num)
            }
            return nil
        }
    }

    return []
}

// Parse integer
public func parseInt(_ s: String) -> Int {
    return Int(s.trimmingCharacters(in: .whitespaces)) ?? 0
}

// Parse double
public func parseDouble(_ s: String) -> Double {
    return Double(s.trimmingCharacters(in: .whitespaces)) ?? 0.0
}

// Parse string (remove quotes)
public func parseString(_ s: String) -> String {
    return s.trimmingCharacters(in: .whitespaces)
            .trimmingCharacters(in: CharacterSet(charactersIn: "\"'"))
}

// Parse boolean
public func parseBool(_ s: String) -> Bool {
    let lower = s.trimmingCharacters(in: .whitespaces).lowercased()
    return lower == "true" || lower == "1" || lower == "yes"
}

// ========================================
// Linked List Operations
// ========================================

// Convert array to linked list
public func arrayToList(_ arr: [Int]) -> ListNode? {
    guard !arr.isEmpty else { return nil }

    let head = ListNode(arr[0])
    var current = head

    for i in 1..<arr.count {
        current.next = ListNode(arr[i])
        current = current.next!
    }

    return head
}

// Convert linked list to array
public func listToArray(_ head: ListNode?) -> [Int] {
    var result: [Int] = []
    var current = head

    while let node = current {
        result.append(node.val)
        current = node.next
    }

    return result
}

// Parse string to linked list
public func parseLinkedList(_ s: String) -> ListNode? {
    return arrayToList(parseIntArray(s))
}

// ========================================
// Binary Tree Operations
// ========================================

// Convert LeetCode format array to binary tree
public func arrayToTree(_ arr: [Int?]) -> TreeNode? {
    guard !arr.isEmpty, let firstVal = arr[0] else { return nil }

    let root = TreeNode(firstVal)
    var queue: [TreeNode] = [root]
    var i = 1

    while !queue.isEmpty && i < arr.count {
        let node = queue.removeFirst()

        // Left child
        if i < arr.count {
            if let val = arr[i] {
                node.left = TreeNode(val)
                queue.append(node.left!)
            }
            i += 1
        }

        // Right child
        if i < arr.count {
            if let val = arr[i] {
                node.right = TreeNode(val)
                queue.append(node.right!)
            }
            i += 1
        }
    }

    return root
}

// Convert binary tree to LeetCode format array
public func treeToArray(_ root: TreeNode?) -> [Int?] {
    guard let root = root else { return [] }

    var result: [Int?] = []
    var queue: [TreeNode?] = [root]

    while !queue.isEmpty {
        let node = queue.removeFirst()

        if let node = node {
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        } else {
            result.append(nil)
        }
    }

    // Remove trailing nils
    while let last = result.last, last == nil {
        result.removeLast()
    }

    return result
}

// Parse string to binary tree
public func parseTree(_ s: String) -> TreeNode? {
    return arrayToTree(parseNullableIntArray(s))
}

// ========================================
// Output Formatting
// ========================================

// Format integer array
public func formatIntArray(_ arr: [Int]) -> String {
    if let data = try? JSONSerialization.data(withJSONObject: arr),
       let str = String(data: data, encoding: .utf8) {
        return str
    }
    return "[\(arr.map { String($0) }.joined(separator: ","))]"
}

// Format string array
public func formatStringArray(_ arr: [String]) -> String {
    if let data = try? JSONSerialization.data(withJSONObject: arr),
       let str = String(data: data, encoding: .utf8) {
        return str
    }
    return "[\(arr.map { "\"\($0)\"" }.joined(separator: ","))]"
}

// Format nullable integer array
public func formatNullableIntArray(_ arr: [Int?]) -> String {
    let mapped: [Any] = arr.map { $0 as Any }
    if let data = try? JSONSerialization.data(withJSONObject: mapped),
       let str = String(data: data, encoding: .utf8) {
        return str.replacingOccurrences(of: "<null>", with: "null")
    }
    return "[\(arr.map { $0 != nil ? String($0!) : "null" }.joined(separator: ","))]"
}

// Format linked list
public func formatList(_ head: ListNode?) -> String {
    return formatIntArray(listToArray(head))
}

// Format binary tree
public func formatTree(_ root: TreeNode?) -> String {
    return formatNullableIntArray(treeToArray(root))
}

// Format boolean
public func formatBool(_ val: Bool) -> String {
    return val ? "true" : "false"
}

// ========================================
// 2D Array Operations
// ========================================

// Parse 2D integer array
public func parseIntMatrix(_ s: String) -> [[Int]] {
    let trimmed = s.trimmingCharacters(in: .whitespaces)
    if trimmed.isEmpty || trimmed == "[]" {
        return []
    }

    if let data = trimmed.data(using: .utf8),
       let arr = try? JSONSerialization.jsonObject(with: data) as? [[Int]] {
        return arr
    }

    return []
}

// Format 2D integer array
public func formatIntMatrix(_ matrix: [[Int]]) -> String {
    if let data = try? JSONSerialization.data(withJSONObject: matrix),
       let str = String(data: data, encoding: .utf8) {
        return str
    }
    return "[]"
}

// ========================================
// LeetCode Input Parsing
// ========================================

// Parse LeetCode format input (e.g., "nums = [1,2,3], target = 9")
public func parseLeetCodeInput(_ input: String) -> [String] {
    var results: [String] = []

    // Split by comma, but handle arrays properly
    var current = ""
    var bracketDepth = 0

    for char in input {
        if char == "[" {
            bracketDepth += 1
            current.append(char)
        } else if char == "]" {
            bracketDepth -= 1
            current.append(char)
        } else if char == "," && bracketDepth == 0 {
            results.append(extractValue(current))
            current = ""
        } else {
            current.append(char)
        }
    }

    if !current.isEmpty {
        results.append(extractValue(current))
    }

    return results
}

// Extract value from "name = value" format
private func extractValue(_ s: String) -> String {
    if let eqIndex = s.firstIndex(of: "=") {
        return String(s[s.index(after: eqIndex)...]).trimmingCharacters(in: .whitespaces)
    }
    return s.trimmingCharacters(in: .whitespaces)
}

// ========================================
// Auto-type Parsing (like Python's parse_input)
// ========================================

// Parse input with automatic type detection
public func parseInput(_ s: String) -> Any {
    let trimmed = s.trimmingCharacters(in: .whitespaces)

    // Empty check
    if trimmed.isEmpty {
        return ""
    }

    // Array detection
    if trimmed.hasPrefix("[") {
        // 2D array detection
        if trimmed.hasPrefix("[[") {
            return parseIntMatrix(trimmed)
        }
        // Check if it's a string array or int array
        if let data = trimmed.data(using: .utf8),
           let arr = try? JSONSerialization.jsonObject(with: data) as? [Any] {
            // Check first element type
            if let first = arr.first {
                if first is String {
                    return parseStringArray(trimmed)
                }
            }
        }
        return parseIntArray(trimmed)
    }

    // Boolean detection
    let lower = trimmed.lowercased()
    if lower == "true" || lower == "false" {
        return parseBool(trimmed)
    }

    // Integer detection
    if let intVal = Int(trimmed) {
        return intVal
    }

    // Double detection
    if let doubleVal = Double(trimmed) {
        return doubleVal
    }

    // Default: string
    return parseString(trimmed)
}

// Parse LeetCode input and return parsed values
public func parseLeetCodeInputs(_ input: String) -> [Any] {
    let stringInputs = parseLeetCodeInput(input)
    return stringInputs.map { parseInput($0) }
}

// ========================================
// Generic Output Formatting
// ========================================

// Format any output value to string
public func formatOutput(_ val: Any) -> String {
    if let arr = val as? [Int] {
        return formatIntArray(arr)
    } else if let arr = val as? [String] {
        return formatStringArray(arr)
    } else if let arr = val as? [[Int]] {
        return formatIntMatrix(arr)
    } else if let arr = val as? [Int?] {
        return formatNullableIntArray(arr)
    } else if let node = val as? ListNode {
        return formatList(node)
    } else if let node = val as? TreeNode {
        return formatTree(node)
    } else if let b = val as? Bool {
        return formatBool(b)
    } else if let n = val as? Int {
        return String(n)
    } else if let d = val as? Double {
        return String(d)
    } else if let s = val as? String {
        return s
    }
    return String(describing: val)
}
